### PR TITLE
Add ifft helpers, basic tests

### LINKED ISSUE #87: ENH ifft functions or ifft mode in our current functions?
In making an example, I found myself requiring an ifft equivalent. This can be achieved with our current functions by flipping the isign and adding a normalization of the output. This is a bit obscure to obtain and maybe we need a shortcut for it

### PR DESCRIPTION
**Pull request by WardBrian**
Closes #87

### CODE REVIEW

File: docs/api.rst [hunk 1]
```diff
@@ -6,3 +6,14 @@ API Reference
 .. autofunction:: pytorch_finufft.functional.finufft_type1
 
 .. autofunction:: pytorch_finufft.functional.finufft_type2
+
+Inverse Transform helper functions
+----------------------------------
+
+Both of these functions are provided merely as helpers, they call
```
Comment (eickenberg) [inline docs/api.rst:13]:
```suggestion
Both of these functions are provided merely as helpers. They call
```

File: pytorch_finufft/functional.py [hunk 2]
```diff
@@ -401,3 +401,34 @@ def finufft_type2(
...
+    points: torch.Tensor,
+    values: torch.Tensor,
+    output_shape: Union[int, Tuple[int], Tuple[int, int], Tuple[int, int, int]],
+    **finufftkwargs: Union[int, float],
+) -> torch.Tensor:
+    """
+    Equivalent to :func:`~pytorch_finufft.functional.finufft_type1`,
+    but passing ``isign=1`` and dividing by the size of the output.
+    """
+    finufftkwargs["isign"] = 1
```
Comment (eickenberg) [inline pytorch_finufft/functional.py:416]: This would be the inverse of the `finufft_type1` with default arguments.
Conceivably, one could also say that this function should flip whatever was put into `finufftkwargs['isign']`.
But that might be a little bit too sketchy in terms of behavior?

Could we maybe warn if `isign` was expclitly set in finufftkwargs to alert the user to the fact that this one won't be used?

File: pytorch_finufft/functional.py [hunk 3]
```diff
@@ -401,3 +401,34 @@ def finufft_type2(
...
+    res: torch.Tensor = finufft_type1(points, values, output_shape, **finufftkwargs)
+    res = res / res.numel()
```
Comment (eickenberg) [inline pytorch_finufft/functional.py:418]: might not hold up with batching. Should probably be `np.prod(res.shape[-points.shape[0]:])`or something similar to `np.prod(res.shape[-ndim:])`. Maybe there is a better way to defined `ndim` here than `points.shape[0]`

File: pytorch_finufft/functional.py [hunk 4]
```diff
@@ -401,3 +401,34 @@ def finufft_type2(
...
+def finuifft_type2(
+    points: torch.Tensor,
+    targets: torch.Tensor,
+    **finufftkwargs: Union[int, float],
+) -> torch.Tensor:
+    """
+    Equivalent to :func:`~pytorch_finufft.functional.finufft_type2`,
+    but passing ``isign=1`` and dividing by the size of the output.
+    """
+    finufftkwargs["isign"] = 1
```
Comment (eickenberg) [inline pytorch_finufft/functional.py:431]: similar concern about the situation where `isign` already exists in finufftkwargs. Does it get flipped?

File: pytorch_finufft/functional.py [hunk 5]
```diff
@@ -401,3 +401,34 @@ def finufft_type2(
...
+    res: torch.Tensor = finufft_type2(points, targets, **finufftkwargs)
+    res = res / res.numel()
```
Comment (eickenberg) [inline pytorch_finufft/functional.py:433]: Similar concern about batch axis making this normalization be off

File: tests/test_inverses.py [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+        values,
+        tuple(N for _ in range(dim)),
+    )
+
+    back = pytorch_finufft.functional.finuifft_type2(
+        points,
+        finufft_out,
+    )
+
+    np.testing.assert_allclose(values.cpu().numpy(), back.cpu().numpy(), atol=1e-4)
```
Comment (eickenberg) [inline tests/test_inverses.py:34]: good test

File: tests/test_inverses.py [hunk 7]
```diff
@@ -0,0 +1,86 @@
+import numpy as np
+import pytest
+import torch
+
+import pytorch_finufft
+
+
+def check_t2_ifft_undoes_t1(N: int, dim: int, device: str) -> None:
```
Comment (eickenberg) [inline tests/test_inverses.py:8]: it strikes me that maybe we need to name `t1_ifft` to be the one that is actually type 2 ...

File: tests/test_inverses.py [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+    100,
+    101,
+]
+
+dims = [1, 2, 3]
+
+
+@pytest.mark.parametrize("N", Ns)
+@pytest.mark.parametrize("dim", dims)
+def test_t2_ifft_undoes_t1_forward_CPU(N, dim):
```
Comment (eickenberg) [inline tests/test_inverses.py:50]: cuda test not necessary?

File: tests/test_inverses.py [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+    check_t2_ifft_undoes_t1(N, dim, "cpu")
+
+
+def check_t1_ifft_undoes_t2(N: int, dim: int, device: str) -> None:
+    """
+    Tests that nuifft_type1 undoes nufft_type2
```
Comment (eickenberg) [inline tests/test_inverses.py:56]: similar Q: should we name the ifft by what type it actually is or what type it is undoing?

**Review by eickenberg — APPROVED** [review]
good to go as is, but some questions about naming and defaults came to mind

File: tests/test_inverses.py (hunk 8 above)
Comment (WardBrian) [reply tests/test_inverses.py:50]: Shouldn't be, any difference between them in this case would certainly be a FINUFFT bug rather than ours. I can add if wanted

File: pytorch_finufft/functional.py (hunk 5 above)
Comment (WardBrian) [reply pytorch_finufft/functional.py:433]: Will revisit after #97 which adds batching, good point

File: pytorch_finufft/functional.py (hunk 2 above)
Comment (WardBrian) [reply pytorch_finufft/functional.py:416]: I think a warning (or even a hard error?) is more appropriate than flipping the input

File: tests/test_inverses.py (hunk 7 above)
Comment (WardBrian) [reply tests/test_inverses.py:8]: Yeah, happy to discuss naming. We could also be clearer in the docstring which function this inverts, not just which one it wraps

**Comment by WardBrian** [comment]
I've made the suggested changes up to the point of a possible re-naming. 

I did have one question @eickenberg, which is that the finufft docs seem to go out of their way to state that they do _not_ provide inverse transforms. See, e,g: https://finufft.readthedocs.io/en/latest/trouble.html#mathematical-issues-and-advice

>The type 1 and type 2 transforms are adjoints but **not inverses of each other** (unlike in the plain FFT case, where, up to a constant factor , the adjoint is the inverse)

or https://finufft.readthedocs.io/en/latest/overview.html#why-finufft-features-and-comparison-against-other-cpu-nufft-libraries

> Note that there are other tasks (eg, transforms on spheres, inverse NUFFTs) provided by other libraries, such as NFFT3, that FINUFFT does not provide.


Are we using 'inverse' in a different sense than they mean?

**Comment by WardBrian** [comment]
Yep, treating the adjoint like the inverse would be bad. Closing this and the related issue
